docs/src/abigen/the-abigen-macro.md

```diff
@@ -0,0 +1,106 @@
+# abigen!
+
+`abigen!` is a procedural macro -- it generates code. It accepts inputs in the format of:
+```text
+ProgramType(name="MyProgramType", abi="my_program-abi.json")...
+```
+where:
+
+`ProgramType` is one of: `Contract`, `Script` or `Predicate`,
+
+`name` is the name that will be given to the generated bindings,
+
+`abi` is either a path to the json abi file or its actual contents.
+
+---
+So, an `abigen!` which generates bindings for two contracts and one script looks like this:
+```rust,ignore
+{{#include ../../../examples/abigen/src/lib.rs:multiple_abigen_program_types}}
+```
+
+## How does the generated code look?
+
+A rough overview:
+```rust,ignore
+pub mod abigen_bindings {
+    pub mod contract_a_mod {
+        struct SomeCustomStruct{/*...*/};
+        // other custom types used in the contract
+        
+        struct ContractA {/*...*/};
+        impl ContractA {/*...*/};
+        // ...
+    }
+    pub mod contract_b_mod {
+        // ...
+    }
+    pub mod my_script_mod {
+        // ...
+    }
+    pub mod my_predicate_mod{
+        // ...
+    }
+    pub mod shared_types{
+        // ...
+    }
+}
+
+pub use contract_a_mod::{/*..*/};
+pub use contract_b_mod::{/*..*/};
+pub use my_predicate_mod::{/*..*/};
+pub use shared_types::{/*..*/};
+```
+
+Each `ProgramType` gets its own `mod` based on the `name` given in the `abigen!`. Inside the respective mods the custom
+types used by that program are generated, as well as the bindings through which the actual calls can be made.
```

Comment:
Not sure if it's an editor thing, but there are a few break lines in the middle of the same paragraphs. Correct me if I'm wrong, but I don't think that was intended.  (I also suggested some clarity-related mods below)

```suggestion
Each `ProgramType` gets its own `mod` based on the `name` given in the `abigen!`. Inside the respective mods, the custom types used by that program are generated, and the bindings through which the actual calls can be made.
```